$ 98 - Interactable Highlight Arrow

Addehman opened the pull request:
Now an arrow appears and hovers over interactable items!

Commit:
$ 98 - Interactable Highlight Arrow

Now an arrow appears and hovers over interactable items!

Commit:
tiny polish on Lever

Addehman commented:
BUG: If the Box is placed beside the Lever, both their InteractArrows won't be removed, and a "MissingReferenceException" pops up. The Interact system needs to prioritize them, so for example it only sees the Box when also the Lever is present. Or, checks which object's center is closer to the player's center.

MrJackey commented on Assets/Scripts/UI/InteractArrow.cs:
```diff
@@ -0,0 +1,30 @@
+	}
+
+	public void BobTheArrow() {
+		if (bobCurrent <= bobStart)
+			isIncreasing = true;
+		else if (bobCurrent >= bobEnd)
+			isIncreasing = false;
+
+		if (isIncreasing) 
+			bobCurrent += bobSpeed * Time.deltaTime;
+		else
+			bobCurrent -= bobSpeed * Time.deltaTime;
```
Why are you using a boolean here? Couldn't you just do the following?
```suggestion
		if (bobCurrent <= bobStart)
			bobCurrent += bobSpeed * Time.deltaTime;
		else if (bobCurrent >= bobEnd)
			bobCurrent -= bobSpeed * Time.deltaTime;
```

MrJackey commented on Assets/Scripts/Player/PlayerHandDetection.cs:
```diff
@@ -3,19 +3,32 @@
 using UnityEngine;
 
 public class PlayerHandDetection : MonoBehaviour {
+	[SerializeField] private GameObject interactArrow;
```
Instead of letting the playerhand monitor the interact arrow, would it be better to let the interactable item handle it themselves?

MrJackey reviewed (changes requested):
Some minor things, I believe only letting the player interact with the closest item should also resolve that bug.

Commit:
Fixed Interaction Selection

It now chooses the item that is closest to the playerHand(GameObject/child).

Only need to consider the distance from the interactable object and the InteractArrow(highlight/selection icon) - should it be higher, like above the player to be more visible, or simply just on top of the interactable object? Maybe even have it adapt to when the player is covering the object to then sit above?

smeas commented on Assets/Scripts/UI/InteractArrow.cs:
```diff
@@ -0,0 +1,34 @@
+
+public class InteractArrow : MonoBehaviour {
+	[SerializeField] private float bobDistance = 0.5f, bobSpeed = 0.1f;
+	//[SerializeField] private AnimationCurve 
```
```suggestion
```

smeas commented on Assets/Scripts/Player/PlayerHandDetection.cs:
```diff
@@ -3,19 +3,72 @@
+	private List<Interactable> interactableList = new List<Interactable>();
+	private GameObject interactArrow;
+	private InteractArrow interactArrowScript;
+	private float lowestDistance = 0;
```
This has no reason to be a field. Just make it a local variable in `SetClosestItem`.

smeas commented on Assets/Scripts/Player/PlayerHandDetection.cs:
```diff
@@ -3,19 +3,72 @@
+		interactArrow.transform.position = closestItem.transform.position + new Vector3(0, closestItem.InteractableArrowHeight, 0);
+		interactArrowScript.Initialize();
+		detectedInteractItem = closestItem;
 	}
```
Cleaned it up a bit.
```suggestion
	private void SetClosestItem() {
		float lowestDistance = float.PositiveInfinity;
		Interactable closestItem = null;

		foreach (Interactable interactable in interactableList) {
			float distance = (transform.position - interactable.transform.position).sqrMagnitude;

			if (distance < lowestDistance) {
				lowestDistance = distance;
				closestItem = interactable;
			}
		}

		if (closestItem == null)
			return;

		interactArrow.transform.position = closestItem.transform.position + new Vector3(0, closestItem.InteractableArrowHeight, 0);
		interactArrowScript.Initialize();
		detectedInteractItem = closestItem;
	}
```

smeas reviewed (changes requested):
The arrow should probably not appear when carrying an item.

smeas committed:
Merge branch 'master' into $98-Interactable-Highlight

Commit:
Applied suggested changes

smeas reviewed (approved):
Nice work!

<img src="https://user-images.githubusercontent.com/8546813/101018772-44168a00-356c-11eb-9fea-3f6cf538d7f5.png" height="120">

MrJackey commented on Assets/Scripts/Player/PlayerHandDetection.cs:
```diff
@@ -3,19 +3,74 @@
+		if (interactable != null) {
+			interactableList.Remove(interactable);	
+			
+			if (interactableList.Count == 0) {
+				detectedInteractItem = null;
+				interactArrow.SetActive(false);
+			}
+			else {
+				detectedInteractItem = null;
+				SetClosestItem();
+			}
```
Can't this be changed?
```suggestion
			detectedInteractItem = null;
			if (interactableList.Count == 0)
				interactArrow.SetActive(false);
```

Commit:
Applied Suggested Changes
